Migrate S.S.C.X509Certificates to new Asn.1 APIs

filipnavara opened the pull request:
Posted to get some early feedback on what is wrong and what should be done differently.

/cc @bartonjs

filipnavara commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/ManagedX509ExtensionProcessor.cs:
```diff
@@ -128,90 +103,67 @@ public virtual void DecodeX509KeyUsageExtension(byte[] encoded, out X509KeyUsage
+            Oid[] keyUsages = AsnSerializer.Deserialize<Oid[]>(encoded, AsnEncodingRules.DER);
+            usages = new OidCollection();
+            foreach (Oid KeyPurposeId in keyUsages)
+                usages.Add(KeyPurposeId);
```
This is the second time I would really appreciate `OidCollection(ICollection<Oid>)` constructor. Oh well.

filipnavara commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/X500NameEncoder.ManagedDecode.cs:
```diff
@@ -135,22 +138,18 @@ internal static partial class X500NameEncoder
 
-            switch (tag)
+            switch ((UniversalTagNumber)tag.TagValue)
             {
```
This cast doesn't feel right.

bartonjs replied:
It's fine if you first check that the TagClass is Universal.

filipnavara commented on src/System.Security.Cryptography.X509Certificates/src/System.Security.Cryptography.X509Certificates.csproj:
```diff
@@ -4,8 +4,12 @@
     <OutputType>Library</OutputType>
     <AssemblyName>System.Security.Cryptography.X509Certificates</AssemblyName>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <NoWarn>CS3016</NoWarn>
```
Taken from Pkcs code. Lot of the code disables this warning using #pragma, but not all. What is the preferred way?

bartonjs replied:
The .csproj way.  The pragmas are older.

filipnavara commented on src/System.Security.Cryptography.X509Certificates/src/System.Security.Cryptography.X509Certificates.csproj:
```diff
@@ -86,6 +107,14 @@
+    <Compile Include="System\Security\Cryptography\X509Certificates\Asn1\TbsCertificateAsn.cs" />
+    <Compile Include="System\Security\Cryptography\X509Certificates\Asn1\ExtensionsAsn.cs" />
+    <Compile Include="System\Security\Cryptography\X509Certificates\Asn1\X509ExtensionAsn.cs" />
+    <Compile Include="System\Security\Cryptography\X509Certificates\Asn1\CertificateAsn.cs" />
```
Not sorted, I know...

filipnavara replied:
... and CertificateAsn, TbsCertificateAsn, ValidityAsn, TimeAsn, ExtensionsAsn and X509ExtensionAsn should go to common code. They are already present in Pkcs (X509ExtensionAsn) or in the common code in simplified version (Certificate, TbsCertificateLite).

filipnavara commented on src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/Asn1/ExtensionsAsn.cs:
```diff
@@ -0,0 +1,18 @@
+    {
+        public X509ExtensionAsn[] Extensions;
+    }
+}
```
Had to introduce this, see issue #31419.

filipnavara commented on src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/CertificateRequest.cs:
```diff
@@ -530,20 +534,92 @@ public X509Certificate2 CreateSelfSigned(DateTimeOffset notBefore, DateTimeOffse
+                }
+                return new ReadOnlyMemory<byte>(serialNumber, leadingZeros, serialNumber.Length - leadingZeros);
+            }
         }
```
This whole methods feels ugly as hell. I thought about declaring the SerialNumber as `BigInteger`, but that only has little-endian constructor and methods, so it feels just as cumbersome.

bartonjs replied:
It has unsigned big-endian ctors and methods in the netcoreapp target.  But this does the work and avoids the extra reading, copying, and writing that BigInteger does.

Since this normalization is a promise of this class (vs rejecting malformed data) this is good.  I prefer this to using BigInteger for the Asn.

bartonjs replied:
```C#
// Did this already check for non-null and non-empty elsewhere?
if (serialNumber[0] == 0 || serialNumber[0] >= 0x80)
{
    // Read it unsigned to stay positive(, man).
    BigInteger serialValue = new BigInteger(serialNumber, isUnsigned: true, isBigEndian: true);
    // Write it signed, so it adds a padding byte if needed
    serialNumber = serialValue.ToByteArray(isUnsigned: false, isBigEndian: true);
}
```

Is the short form (for builds which can use the new BigInteger API).  It's just more expensive (always creates a new output array, creates an intermediate `uint[]` if the parsed value wouldn't fit in an `int`).  Granted, this code isn't that perf-sensitive.

filipnavara replied:
Good catch with the negative number. I will write unit test for it.

bartonjs commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/CertificateData.cs:
```diff
@@ -59,126 +61,57 @@ internal CertificateData(byte[] rawData)
                 throw new CryptographicException();
 
-            SerialNumber = tbsCertificate.ReadIntegerBytes();
+            SerialNumber = tbsCertificate.SerialNumber.ToArray();
```
Why? Just store the TbsCertificateAsn and do the ToArray when asked.

filipnavara replied:
Actually the whole `CertificateData` class could basically be replaced with `CertificateAsn` and few added caches for Issuer and Subject stored as `X500DistinguishedName`. I kept it around only temporarily.

bartonjs commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/CertificateData.cs:
```diff
@@ -59,126 +61,57 @@ internal CertificateData(byte[] rawData)
+            PublicKey = tbsCertificate.SubjectPublicKeyInfo.SubjectPublicKey.ToArray();
+
+            IssuerUniqueId = Version >= 1 ? tbsCertificate.IssuerUniqueId?.ToArray() : null;
+            SubjectUniqueId = Version >= 1 ? tbsCertificate.SubjectUniqueId?.ToArray() : null;
```
Most of the fields should be dropped in favor of saving the serializer output.  These two would maintain consistency with current code by saying that if `Version < 1` and they aren't `null` to throw the invalid encoding exception.

filipnavara replied:
Agreed, in fact the whole class should be dropped (see comment above) in favor of using `CertificateAsn` directly. It doesn't offer any additional value except few checks and cached Issuer+Subject.

bartonjs commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/CertificateData.cs:
```diff
@@ -59,126 +61,57 @@ internal CertificateData(byte[] rawData)
-            if (Version > 1 &&
-                tbsCertificate.HasData &&
-                tbsCertificate.PeekTag() == DerSequenceReader.ContextSpecificConstructedTag3)
+            if (Version >= 2 && tbsCertificate.Extensions.HasValue)
```
This if says to only copy the extensions if the version is 2 or more, and ignore them for 1 or 0.  It lost the "throw if extensions appear in version 0 or 1" part.

Since the whole structure was already parsed (as well as the original code did) this also can be deferred until something asks about it, to save memory.

filipnavara replied:
Good catch.

bartonjs commented on src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/Asn1/TbsCertificateAsn.cs:
```diff
@@ -0,0 +1,65 @@
+        [BitString]
+        public ReadOnlyMemory<byte>? IssuerUniqueId;
+
+        [ExpectedTag(2, ExplicitTag = true)]
```
This field shouldn't be marked as explicit.  The next one should, which probably eliminates the intermediate struct.

filipnavara replied:
Uh, cut and paste error... I will try it when I get back to my computer.

bartonjs commented on src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/CertificateRequest.cs:
```diff
@@ -530,20 +534,92 @@ public X509Certificate2 CreateSelfSigned(DateTimeOffset notBefore, DateTimeOffse
+                foreach (X509Extension extension in CertificateExtensions)
+                {
+                    if (extension == null)
+                        continue;
```
Since this if is not in the top method block it needs explicit braces.

bartonjs commented on src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/TbsCertificate.cs:
```diff
@@ -1,194 +0,0 @@
-            if (utcValue.Year < 1950)
-            {
-                throw new ArgumentOutOfRangeException(propertyName, utcValue, SR.Cryptography_CertReq_DateTooOld);
-            }
```
Was this exceptional state lost?  Or is it a repeat of non-modified code?

filipnavara replied:
It was lost, now it allows years below 1950 and relies on the checks in AsnWriter to do the right thing.

filipnavara replied:
...but it looks like AsnWriter doesn't do checks for negative years.

T-REC-X.680-201508 sec 46 says that GeneralizedTime should be written using "ISO 8601, 4.1.2.2 – Basic format". It doesn't specify any additional restrictions. The 4.1.2.2 representation doesn't allow for negative years as far as I can tell (the expanded representation has to be used). I guess `WriteGeneralizedTimeCore` method should be modified to reject years < 0 just as it rejects years > 9999.

filipnavara replied:
Ah, it doesn't need to check for negative years since DateTime can't represent them.

filipnavara commented:
@bartonjs  Btw, the unit tests in Pkcs fail because there's non-canonical encoding of Key Usage used ('03 02 00 80'). It's easy to modify the X509 code to accept it, but it's not valid according to RFC 5280:

```
Extension  ::=  SEQUENCE  {
     extnID      OBJECT IDENTIFIER,
     critical    BOOLEAN DEFAULT FALSE,
     extnValue   OCTET STRING
                 -- contains the DER encoding of an ASN.1 value
                 -- corresponding to the extension type identified
                 -- by extnID
     }
```

Should that be accepted for compatibility or should it be rejected and the test case fixed?

bartonjs commented:
Based on the variable names in the test, I'm pretty sure that the value used was generated by asking Windows to write a PFX for an ECDSA cert+key, so there might be a compat problem.

My guess is that the encoding of this value inherited the BER context of PKCS8 and the optimization of "I don't need to set the number of unused bits" was taken (e.g. "X.690 11.2.2 doesn't apply").

The Windows reader here calls `CryptDecodeObject` with (BER | DER) (or, rather, (PKCS | X509)), so that's probably why it's passing.  Alternatively, Windows might just be reading it as a BIT STRING instead of a NamedBitList, meaning it doesn't care that the unused bit count value is correct (other than no "unused" bits are set).  That might be a viable workaround, something like

```C#
// Use BIT STRING instead of NamedBitList for compatibility with encodings that neglected
// to assign unusedBitCount
Span<X509KeyUsageFlags> usageSpan = stackalloc X509KeyUsageFlags[1];

if (!reader.TryCopyBitStringBytes(usageSpan.AsBytes(), out _, out _))
{
    throw corrupt;
}

if (!BitConverter.IsLittleEndian)
{
    // flip the bytes
}

return usageSpan[0];
```

filipnavara commented:
I think the easier workaround is just to specify BER for the reader and keep DER for the writer. That should get around exactly that one unused bit check.

bartonjs commented:
Changing to BER would allow `23 04 03 02 00 80`.  If that works for .NET Framework, that's fine.  It would also allow `03 82 00 02 00 80`.  If all of those work, great.  If not, you can't use BER.

My gut says that the snippet I gave would work, it's the reason that Windows picked the values they did.

filipnavara commented:
```
using System;
using System.Security.Cryptography;
using System.Security.Cryptography.X509Certificates;
					
public class Program
{
	public static void Main()
	{
		var ext = new X509KeyUsageExtension(new AsnEncodedData(new byte[] { 0x23, 0x04, 0x03, 0x02, 0x00, 0x80 }), false);
		Console.WriteLine(ext.KeyUsages);
	}
}
```

produces `DigitalSignature` on NetFX. That pretty much confirms that it accepts BER on input.

filipnavara commented:
`EncodeX509KeyUsageExtension` is probably fine. The code in NetFX uses similar construct and produces the correct DER output for `new X509KeyUsageExtension(X509KeyUsageFlags.DigitalSignature, false).RawData`.

bartonjs commented on src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/CertificateRequest.cs:
```diff
@@ -530,20 +534,96 @@ public X509Certificate2 CreateSelfSigned(DateTimeOffset notBefore, DateTimeOffse
+            {
+                // Strip any unnecessary zeros from the beginning.
+                int leadingZeros = 0;
+                while (leadingZeros < serialNumber.Length - 1 && serialNumber[leadingZeros] == 0 && serialNumber[leadingZeros + 1] < 0x80)
```
I think the last clause will ArrayIndexOutOfBounds for all-zeros.

filipnavara replied:
The first clause should prevent that. Then again, it definitely feels hard to read.

filipnavara commented on src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/Asn1/X501AttributeAsn.cs:
```diff
@@ -0,0 +1,39 @@
+        {
+            [AnyValue]
+            internal ReadOnlyMemory<byte>? AttrValue;
+        }
```
This feels like quite a cumbersome way to encode it. I need a `SET` with one element that is exactly the value of `RawData` of the `X501Attribute`.

bartonjs commented on src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/Asn1/TimeAsn.cs:
```diff
@@ -0,0 +1,43 @@
+
+namespace System.Security.Cryptography.X509Certificates.Asn1
+{
+    // Time ::= CHOICE {
```
Missing source URL

filipnavara commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/ManagedX509ExtensionProcessor.cs:
```diff
@@ -63,47 +59,26 @@ public virtual void DecodeX509KeyUsageExtension(byte[] encoded, out X509KeyUsage
             if (hasPathLengthConstraint)
+                constraints.PathLengthConstraint = pathLengthConstraint;
+
+            using (AsnWriter writer = AsnSerializer.Serialize(constraints, AsnEncodingRules.BER))
```
Uh, mistake.

bartonjs commented on src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/Asn1/ValidityAsn.cs:
```diff
@@ -0,0 +1,26 @@
+
+namespace System.Security.Cryptography.X509Certificates.Asn1
+{
+    // Validity ::= SEQUENCE {
```
Missing source URL

filipnavara commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/ManagedX509ExtensionProcessor.cs:
```diff
@@ -231,67 +188,20 @@ public virtual byte[] ComputeCapiSha1OfPublicKey(PublicKey key)
+            spki.Algorithm = new AlgorithmIdentifierAsn { Algorithm = key.Oid, Parameters = key.EncodedParameters.RawData };
+            spki.SubjectPublicKey = key.EncodedKeyValue.RawData;
 
+            using (AsnWriter writer = AsnSerializer.Serialize(spki, AsnEncodingRules.BER))
```
Again, mistake :)

bartonjs commented on src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/Asn1/X501AttributeAsn.cs:
```diff
@@ -0,0 +1,39 @@
+namespace System.Security.Cryptography.X509Certificates.Asn1
+{
+    [StructLayout(LayoutKind.Sequential)]
+    internal struct X501AttributeAsn
```
Missing source URL and ASN.1

bartonjs commented on src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/Asn1/X509ExtensionAsn.cs:
```diff
@@ -0,0 +1,36 @@
+namespace System.Security.Cryptography.X509Certificates.Asn1
+{
+    [StructLayout(LayoutKind.Sequential)]
+    internal struct X509ExtensionAsn
```
Missing source URL and ASN.1

filipnavara replied:
This was actually taken from the existing Pkcs code. I have another PR that moves the class over, so I will fix this after that one is merged and this one rebased.

bartonjs commented on src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/Pkcs10CertificationRequestInfo.cs:
```diff
@@ -36,52 +37,46 @@ internal class Pkcs10CertificationRequestInfo
+                Version = 0,
+                Subject = this.Subject.RawData,
+                SubjectPublicKeyInfo = spki,
+                Attributes = Attributes.Select(e => new X501AttributeAsn(e)).ToArray()
```
I notice that all of your multi-line object initializers omit the optional trailing comma.  We don't have a style rule one way or another, but my personal preference is to always write trailing commas (when multiline), because it cuts down on noise in future edits.

filipnavara replied:
Fine, I will add the commas.

bartonjs commented on src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/Pkcs9ExtensionRequest.cs:
```diff
@@ -19,34 +22,12 @@ private static byte[] EncodeAttribute(IEnumerable<X509Extension> extensions)
-                    continue;
-
-                encodedExtensions.Add(extension.SegmentedEncodedX509Extension());
+                return writer.Encode();
```
I'd write this as a foreach and avoid the double (or triple) copy in the LINQ statement.

```C#
using (AsnWriter writer = new AsnWriter(AsnEncodingRules.DER))
{
    writer.PushSequence();

    foreach (X509Extension e in extensions)
    {
        AsnSerializer.Serialize(new X509ExtensionAsn(e), writer);
    }

    writer.PopSequence();
    return writer.Encode();
}
```

bartonjs commented on src/System.Security.Cryptography.X509Certificates/tests/CertificateCreation/CertificateRequestUsageTests.cs:
```diff
@@ -296,8 +296,11 @@ public static void SelfSign_ECC_DiminishedPoint_UseCertKeys()
+        [InlineData("80")]
+        [InlineData("0080")]
+        [InlineData("00000080")]
+        public static void SerialNumber_AlwaysPositive(string desiredSerial)
```
If there's not already a test for "00000000" (or whatever your favorite all-zeros representation is), please add one, since that was a point of concern from a code review comment.

filipnavara commented:
This is getting really big and somewhat messy to review. What is the best way to proceed?

1) Continue piling stuff into this PR until all DerSequenceReader usages are gone?
2) Wrap this one up, review, fix and merge it and then start a new one(s) for the remaining parts?
3) Split this one into smaller, reviewable, chunks?

My estimate is that around 50% of the code is now converted to the new APIs.

Aside from the review comments above there's one thing I am still concerned about. I believe all `ManagedX509ExtensionProcessor` _decode_ methods should be switched to allow BER encoding. The Windows PAL uses the Crypto API with flags that allow it. I am not sure if I will be able to produce test cases for everything though.

bartonjs commented:
I was going to similarly suggest surprise that additional things keep being added.

I don't have a strong preference for wrapping this up as "phase 1" vs splitting it off into tangible named chunks.  Each has their own benefits.

filipnavara commented:
If it makes the reviews easier I am fine with splitting it up [despite the additional effort on my part].

bartonjs commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/CertificateData.cs:
```diff
@@ -31,156 +33,75 @@ internal struct CertificateData
+        
+        internal byte[] SerialNumber => certificate.TbsCertificate.SerialNumber.ToArray();
+        
+        internal AlgorithmIdentifier TbsSignature => new AlgorithmIdentifier(certificate.TbsCertificate.SignatureAlgorithm);
```
Are all of these properties used?  I think TbsSignature was only built out to ensure we read the values, now the serializer makes sure that's done.

filipnavara replied:
Looks like the remaining ones are used. I will remove this one.

bartonjs commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/ManagedX509ExtensionProcessor.cs:
```diff
@@ -221,8 +173,13 @@ public virtual void DecodeX509SubjectKeyIdentifierExtension(byte[] encoded, out
+            {
+                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+            }
+            return contents.ToArray();
```
reader.ThrowIfNotEmpty() ?

filipnavara replied:
Yep, of course. Ironically I just fixed the same error in Pkcs in the other PR for exactly the same structure :-)

bartonjs commented on src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/Asn1/TbsCertificateAsn.cs:
```diff
@@ -0,0 +1,90 @@
+    //                          -- If present, version MUST be v2 or v3
+    //     subjectUniqueID [2]  IMPLICIT UniqueIdentifier OPTIONAL,
+    //                          -- If present, version MUST be v2 or v3
+    //      extensions      [3]  Extensions OPTIONAL
```
Super-nit: Looks like an extra space got pasted in before `extension`

filipnavara replied:
Removed.

bartonjs commented on src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/Asn1/RSAPssParamAsn.cs:
```diff
@@ -0,0 +1,51 @@
+    //  * don't support the MGF PRF being different than hashAlgorithm
+    //  * use saltLength==hashLength
+    //  * don't allow custom trailer
+    // we don't have to worry about any of the DEFAULTs. (specify, specify, specify, omit).
```
This comment doesn't make sense now that this is visible outside that one consumer.

Since this could now get easily picked up by some other component, please specify the DefaultValue values for `hashAlgorithm` and `maskGenAlgorithm`

`hashAlgorithm`'s default is (I think) `30 09 06 05 2B 0E 03 02 1A 05 00` and `maskGenAlgorithm`'s is `30 16 06 09 2A 86 48 86 F7 0D 01 01 08 30 09 06 05 2B 0E 03 02 1A 05 00`.

At least, based on the comments on lines 19 and 20.

The comment for lines 23-25 seem to still belong at the caller, about how to build the AlgorithmIdentifierAsn.

filipnavara replied:
I think the default values are just 30 07 06 05 2B 0E 03 02 1A and 30 14 06 09 2A 86 48 86 F7 0D 01 01 08 30 07 06 05 2B 0E 03 02 1A (without the NULLs at the end).

bartonjs replied:
Omitting the NULL is correct for the SHA-2 algorithms, but https://tools.ietf.org/html/rfc3447 says

```ASN.1
sha1    HashAlgorithm ::= {
    algorithm   id-sha1,
    parameters  SHA1Parameters : NULL  -- included for compatibility
                                       -- with existing implementations
}

mgf1SHA1    MaskGenAlgorithm ::= {
    algorithm   id-mgf1,
    parameters  HashAlgorithm : sha1
}

--
-- AlgorithmIdentifier.parameters for id-RSASSA-PSS.
-- Note that the tags in this Sequence are explicit.
--
RSASSA-PSS-params ::= SEQUENCE {
    hashAlgorithm      [0] HashAlgorithm      DEFAULT sha1,
    maskGenAlgorithm   [1] MaskGenAlgorithm   DEFAULT mgf1SHA1,
    saltLength         [2] INTEGER            DEFAULT 20,
    trailerField       [3] TrailerField       DEFAULT trailerFieldBC
}
```

So the NULLs are semantically within the SHA-1-based default values.

bartonjs replied:
I based it on the default value for Version in TbsCertificate.  Looking at the serializer code I think it does.

I don't see any specific tests for serializing and deserializing default explicit values in the encoding tests library, seems like it would be good to add one so we can say for certain :smile:.

filipnavara replied:
The default value of Version in TbsCertificate is currently not used in practice or any of the tests afaik (and it was copied from the older Pkcs code). The default values of RSAPssParamAsn for the last two fields are tested to some extent though (`RSAPssX509SignatureGeneratorTests.SignatureAlgorithm_Encoding`). I will make a test for it.

bartonjs replied:
> I will make a test for it.

Yay :smile:.

I second-, third-, fourth-, and fifth- guessed myself as to whether the default values reading and writing was inside or outside the explicit tag.

filipnavara replied:
Interesting, there's one more copy of `TbsCertificate` in `X509DeserializeTests` and that one has the `DefaultValue` without the explicit tag. :-) It still doesn't use the default value in the tests though.

bartonjs replied:
That would explain why I kept n-th guessing myself.  It probably just hasn't come up in a read/write scenario yet.

The smoothest approach is probably to fix that aspect of the serializer (with tests) in a separate PR, and then resume this.  Otherwise, make sure that nothing is doing both reading and writing of the same structure which has default+explicit fields.  (AppleCertificatePal is reading TbsCertificate.Version, CertificateRequest is writing it, so they can't both go in together)

filipnavara replied:
I will do more tests tomorrow to see if I am able to fix it.

filipnavara replied:
Ok, I submitted PR for the deserializer along with unit tests. Now I noticed that there's already PssParamsAsn in the Pkcs code. Will reuse that one after I rebase this PR :-)

filipnavara replied:
...ouch, but that one also has the default values without the explicit NULL at the end.

bartonjs commented on src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/CertificateRequest.cs:
```diff
@@ -2,8 +2,12 @@
+using System.Collections.Generic;
 using System.Collections.ObjectModel;
 using System.Diagnostics;
+using System.Linq;
```
Is anything still using System.Linq?

bartonjs commented on src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/Pkcs9ExtensionRequest.cs:
```diff
@@ -3,6 +3,9 @@
 // See the LICENSE file in the project root for more information.
 
 using System.Collections.Generic;
+using System.Linq;
```
I don't think this using-import statement is needed anymore.

filipnavara commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/ManagedX509ExtensionProcessor.cs:
```diff
@@ -15,30 +16,25 @@ internal class ManagedX509ExtensionProcessor
-            return ConcatenateArrays(segments);
+            using (AsnWriter writer = new AsnWriter(AsnEncodingRules.DER))
+            {
+                writer.WriteNamedBitList<KeyUsageFlagsAsn>(keyUsagesAsn);
```
Removed.

filipnavara commented on src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/Pkcs10CertificationRequestInfo.cs:
```diff
@@ -36,52 +37,46 @@ internal class Pkcs10CertificationRequestInfo
+            AlgorithmIdentifierAsn signatureAlgorithmAsn;
+
+            // Deserialization also does validation of the value.
+            signatureAlgorithmAsn = AsnSerializer.Deserialize<AlgorithmIdentifierAsn>(signatureAlgorithm, AsnEncodingRules.DER);
```
Hmm, are you saying that the deserialization of ANY is weaker check than `DerSequenceReader.ValidateAndSkipDerValue`?

bartonjs replied:
Yep.  Deserializing an ANY stops recursion.  ValidateAndSkipDerValue recurs on constructed values.

bartonjs commented:
FYI, https://github.com/dotnet/corefx/pull/31344 will conflict with this.  Since it's (ironically, since it's not writing any different code) harder to review I'm going to let it go in first.

filipnavara commented:
No problem, I will rebase this after the other PR(s) go in.

filipnavara commented:
I'll rebase this once #31470 and #31502 goes in and drop the WIP tag.

bartonjs commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/ManagedX509ExtensionProcessor.cs:
```diff
@@ -128,90 +103,67 @@ public virtual void DecodeX509KeyUsageExtension(byte[] encoded, out X509KeyUsage
-                pathLengthConstraint = 0;
-                hasPathLengthConstraint = false;
-            }
+            BasicConstraintsAsn constraints = AsnSerializer.Deserialize<BasicConstraintsAsn>(encoded, AsnEncodingRules.BER);
```
For all of the places where the deserialization is being done in BER (and it is currently DER in master because it uses DerSequenceReader) please add a test using a BER-but-not-DER value, so that tests don't pass if this is flipped to DER in a later change.

I know we already have one for KeyUsage, but I don't think BasicConstraints or SubjectKeyIdentifier gained one.

filipnavara replied:
Yeah, I didn't write the tests yet. Will do it asap.

filipnavara replied:
Done

bartonjs commented on src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/Asn1/KeyUsageFlagsAsn.cs:
```diff
@@ -0,0 +1,26 @@
+
+namespace System.Security.Cryptography.X509Certificates.Asn1
+{
+    /// <summary>Version of X509KeyUsageFlags with reversed bit order.</summary>
```
Missing ASN (https://tools.ietf.org/html/rfc3280#section-4.2.1.3)

```asn.1
      KeyUsage ::= BIT STRING {
           digitalSignature        (0),
           nonRepudiation          (1),
           keyEncipherment         (2),
           dataEncipherment        (3),
           keyAgreement            (4),
           keyCertSign             (5),
           cRLSign                 (6),
           encipherOnly            (7),
           decipherOnly            (8) }
```

Then I'd define the C# enum in "the natural way":

```C#
    [Flags]
    internal enum KeyUsageFlagsAsn
    {
        None = 0,
        DigitalSignature = 1 << 0,
        NonRepudiation = 1 << 1,
        KeyEncipherment = 1 << 2,
        DataEncipherment = 1 << 3,
        KeyAgreement = 1 << 4,
        KeyCertSign = 1 << 5,
        CrlSign = 1 << 6,
        EncipherOnly = 1 << 7,
        DecipherOnly = 1 << 8,
    }
```

filipnavara commented:
Once this gets in I will submit phase 2 that removes all remaining `DerEncoder` usages.

bartonjs commented:
> Once this gets in I will submit phase 2 that removes all remaining DerEncoder usages.

How much is left? Mainly I'm thinking about reviewability.

filipnavara commented:
@bartonjs I will split the rest in chunks that are easier to review. I have last few small chunks to cleanup and test, but other than that it's only matter of splitting the code into self-contained PRs.

jkotas commented:
@bartonjs The new Asn APIs are doing reflection-based serialization. A simple https request is going a bunch of heavy duty reflection as a result after this change. Do we understand the performance characteristics of the new reflection-based ASN serializer? The old one seemed to be lean and fast. The new one is exact opposite.

filipnavara commented:
@jkotas Do you know about some specific HTTPS code path that ends up using the serializer? I can imagine few that are hit on macOS (CertificateData class), but I would be interested in knowing some hot paths that need to be optimized.

I cannot answer about the performance characteristics of the serializer, but I was concerned about it as well. I came to the conclusion that it would be trivial change to allow caching of the serializer that is built using reflection APIs. That may not be solution to all scenarios, but it could help substantially for APIs that are called repeatedly.

jkotas commented:
> macOS 

Yes, somebody hit on macOS (https://github.com/dotnet/corert/issues/6221)

>   trivial change to allow caching of the serializer that is built using reflection APIs. That may not be solution to all scenarios

Right. That is the usual path that the reflection-based serializers take. They start as nice easy to follow code. Then it is found that they have slow throughput and caching schemes are invented for them that makes them complex. And then it is found that they have slow startup and the fix is to generated as code at build time. 

It would recommend to generate them as code at build time in the first place and not have the reflection path at all. It would save us from going through this path yet again.

filipnavara commented:
I totally get what you mean. We have been working with this problematic lifetime flow of serializers since many years ago...and XML serializers invoking C# compiler at runtime in NetFX was one hell for us for many many years. 

I'm open to any ideas or solutions if they are properly discussed and agreed upon. Note that the ASN code is already powering S.S.C.Pkcs and some S.S.C.Algorithms code, so this change merely exposed the problem on a more used code path.

The caching is really trivial to add since the design of AsnSerializer already builds the deserializer first and then uses it to do the deserialization once and throws it away. It doesn't solve the problem of having to have the reflection metadata available and startup time.
